### PR TITLE
BUG: Index.join with different names doesn't return None name

### LINKED ISSUE #57065: BUG: Index.join preserving names incorrectly from pandas-2.x
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [2]: import pandas as pd
In [3]: pd.__version__
Out[3]: '2.2.0'
In [4]: idx1 = pd.Index([1, 1, 2, 3, 4, 4], name='a')
In [5]: idx2 = pd.Index([2, 5, 3, 4], name='b')
In [6]: idx1.join(idx2)
Out[6]: Index([1, 1, 2, 3, 4, 4], dtype='int64', name='a')
In [7]: idx2.join(idx1)
Out[7]: Index([2, 5, 3, 4, 4], dtype='int64', name='b')
```

### Issue Description

When two index objects have different names, in pandas-1.x versions the end result index would have `None` as index name. Now in pandas-2.x the result index is always preserving the existing index name it is being called upon.

### Expected Behavior

```python
In [2]: import pandas as pd
In [3]: pd.__version__
In [4]: idx1 = pd.Index([1, 1, 2, 3, 4, 4], name='a')
In [5]: idx2 = pd.Index([2, 5, 3, 4], name='b')
In [6]: idx1.join(idx2)
Out[6]: Index([1, 1, 2, 3, 4, 4], dtype='int64')
In [7]: idx2.join(idx1)
Out[7]: Index([2, 5, 3, 4, 4], dtype='int64')
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : fd3f57170aa1af588ba877e8e28c158a20a4886d
python                : 3.10.2.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.2.0
Version               : Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : en_US.UTF-8
pandas                : 2.2.0
numpy                 : 1.23.2
pytz                  : 2022.2.1
dateutil              : 2.8.2
setuptools            : 60.2.0
pip                   : 21.3.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : 8.5.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : 2.8.4
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 11.0.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [ ] closes #57065 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by lukemanley** [comment]
I think this introduces an inconsistency between the non-unique and monotonic join paths.

xref https://github.com/pandas-dev/pandas/issues/55815 where @rhshadrach and I tried to determine the intended behavior. 

When left/right names differ, the choices seem to be:
- set name to none
- set name to caller (left)
- use name of caller (left) in all cases except `how="right"`

As of 2.2.0 I think there are cases of 1 and 3 occuring in various places. #55815 attempted to implement #3 consistently but maybe it should be #1 as you're doing in this PR.

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -116,9 +116,6 @@ Performance improvements
 
 Bug fixes
 ~~~~~~~~~
-- Fixed bug in :meth:`DataFrame.join` inconsistently setting result index name (:issue:`55815`)
-- Fixed bug in :meth:`Series.diff` allowing non-integer values for the ``periods`` argument. (:issue:`56607`)
```
Comment (lukemanley) [inline doc/source/whatsnew/v3.0.0.rst:120]: did you intend to remove the `Series.diff` line?

Comment (lukemanley) [reply doc/source/whatsnew/v3.0.0.rst:120]: nevermind, I see you moved it

**Comment by mroeschke** [comment]
Thanks for the reference @lukemanley 

For `set name to caller (left) in all cases except how="right"`, is there any case where "caller" is "ambiguous" with a function based join like `pd.concat` or `pd.merge`? Should that apply to arithmetic operations too idx1 + idx2 ~ `idx1.__add__(idx2)`?

I slightly prefer the return None approach as it's already being done in arithmetic cases and appears simpler, but not sure if there's a use case where joins specifically should keep names.

**Comment by rhshadrach** [comment]
> I slightly prefer the return None approach as it's already being done in arithmetic cases and appears simpler, but not sure if there's a use case where joins specifically should keep names.

Are we meaning to have rules for Index vs Series?

```
idx1 = Index([1, 1, 2, 3], name="a")
idx2 = Index([2, 5, 3, 4], name="b")

ser1 = pd.Series(1, index=idx1)
ser2 = pd.Series(2, index=idx2)

print(idx1+idx2)
# Index([3, 6, 5, 7], dtype='int64')

print(ser1+ser2)
# a
# 1    NaN
# 1    NaN
# 2    3.0
# 3    3.0
# 4    NaN
# 5    NaN
# dtype: float64
```

I can't say I've ever directly used `Index.join`, only `DataFrame.join`. Likewise with arithmetic. But if there isn't a good reason why these two behaviors should diverge (not sure), I would think users expect them to match.

**Comment by lukemanley** [comment]
> I slightly prefer the return None approach as it's already being done in arithmetic cases and appears simpler, but not sure if there's a use case where joins specifically should keep names.

Keeping consistency with implicit alignment (e.g. arithmetic cases) seems reasonable to me.

**Comment by mroeschke** [comment]
OK with merging here?

**Comment by lukemanley** [comment]
> OK with merging here?

It looks like this only impacts the `_join_non_unique` path. Doesn't this create an inconsistency with `_join_monotonic`? For example, what would this return:

```
import pandas as pd

a = pd.Index([1, 2, 3], name="a")
b = pd.Index([2, 3, 4], name="b")

a.join(b, how="inner")
```

**Comment by rhshadrach** [comment]
> OK with merging here?

I think it'd be good to decide if we are okay with the discrepancy between Index and Series behavior (https://github.com/pandas-dev/pandas/pull/57074#issuecomment-1912579392) that this is further widening. It seems to me changing Series behavior would be quite disruptive, while I personally never do arithmetic with index objects.

**Comment by mroeschke** [comment]
I guess this needs more discussion first so I'll close for now

**Comment by rhshadrach** [comment]
@mroeschke - any thoughts on the discrepancy between the resulting name of `idx1 + idx2` and `ser1 + ser2` in https://github.com/pandas-dev/pandas/pull/57074#issuecomment-1912579392?

**Comment by mroeschke** [comment]
I also think there shouldn't be a result-name discrepancy between `idx1 + idx2` and `ser1 + ser2`, but namely `get_op_result_name` should also be used in both these cases

**Pull request closed without merging**